src/ast/passes/resolve_imports.cpp

```diff
@@ -1,29 +1,242 @@
+                                        const std::filesystem::path &path)
+{
+  auto [_, added] = imports.objects.emplace(name, ExternalObject(path));
+  assert(added);
+  return OK();
+}
+
+Result<OK> ResolveImports::importBitcode([[maybe_unused]] Node &node,
+                                         const std::string &name,
+                                         const std::string &contents)
+{
+  auto [_, added] = imports.objects.emplace(name, Bitcode(contents));
+  assert(added);
+  return OK();
+}
+
+Result<OK> ResolveImports::importAny(Node &node,
+                                     const std::string &name,
+                                     const std::filesystem::path &path)
+{
+  if (imports.objects.contains(name)) {
+    return OK(); // Already added.
+  }
+
+  if (std::filesystem::is_directory(path)) {
+    // Recursively import all entries in the directory.
+    for (const auto &entry : std::filesystem::directory_iterator(path)) {
+      auto ok = importAny(node,
+                          name + "/" + entry.path().filename().string(),
+                          entry.path());
+      if (!ok) {
+        return ok.takeError();
+      }
+    }
+    return OK();
+  } else {
+    // Import any support file-based extensions.
+    if (path.extension() == ".bt") {
+      return importNative(node, name, path);
+    } else if (path.extension() == ".o") {
+      return importObject(node, name, path);
+    } else {
+      node.addError() << "unknown import type: " << path.filename();
+      return OK();
+    }
+  }
+}
+
+Result<OK> ResolveImports::importAny(Node &node,
+                                     const std::string &name,
+                                     const std::string &data)
+{
+  if (imports.objects.contains(name)) {
+    return OK(); // Already added.
+  }
+
+  // Import supported extensions.
+  std::filesystem::path path(name);
+  if (path.extension() == ".bt") {
+    return importNative(node, name, data);
+  } else if (path.extension() == ".bc") {
+    return importBitcode(node, name, data);
+  } else if (path.extension() == ".btf") {
+    // Ignore for now: these are BTF types.
+    return OK();
+  } else {
+    node.addError() << "unknown import type: " << path;
+    return OK();
+  }
+}
+
+Result<OK> ResolveImports::importAny(Node &node, const std::string &name)
+{
+  if (imports.objects.contains(name)) {
+    return OK(); // Already added.
+  }
+
+  std::vector<std::string> checked;
+  if (allow_paths) {
+    for (const auto &import_path : paths_) {
+      // Check to see if this specific file exists.
+      auto path = import_path / name;
+      checked.emplace_back(path.string());
+      if (!std::filesystem::exists(path)) {
+        continue; // No file found.
+      }
+
+      // Attempt the import.
+      return importAny(node, name, path);
+    }
+  }
+
+  // See if this matches a set of builtins. Note that we do the "directory"
+  // expansion here, importing anything that is matching as a path.
```

Comment:
Should this be inserted into `checked`?

Reply:
This is the set of files that are effectively built in to the binary itself, so there's nothing under the user control. I thought about including the list of builtins, but I think we would want to surface that in another way.

Reply:
I was thinking more if someone tries to include the standard library but typos, the following error would be a good hint about the correct spelling.

Reply:
I've added `utils/similar.h`, which includes a helper to give a "similarly score" for two strings. If they are similar, they will be included and added as a hint. To do this, I also added support for multiple hints.

Reply:
> I was thinking more if someone tries to include the standard library but typos

Are we not including the standard library by default for all scripts? Or am I confused.

Reply:
Correctly -- if it is included implicitly, then it will always be sourced from the embedded version. If users import explicitly, then it will do the standard file system search (same as used for all explicit imports).